add-ons/updating/tb78/convert-legacy.md

```diff
@@ -0,0 +1,146 @@
+
+Option windows are replaced by [settings pages](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Implement_a_settings_page) defined in `manifest.json`:
+```
+"options_ui": {
+  "page": "options.html"
+},
+```
+Instead of an XUL dialog, use the specified HTML document will be accessible to the user through the Add-on manager. From that document, all WebExtension and MailExtension APIs can be accessed in the same way as form a background script. The settings themselves should be stored using one of the new APIs to store data, such as [`storage`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/storage), so it may be necessary to add an experiment to migrate existing settings from nsIPrefBranch or other mechanisms not accessible through modern APIs.
+
+
+## Replacing chrome.manifest
+
+The `chrome.manifest` file is no longer supported. Many mechanisms have a more or less direct equivalent in the WebExtension world:
+* **interfaces** it should no longer be necessary to use binary interfaces, as binary components are deprecated for a long time. To access custom methods on JS-implemented classes from an experiment, use `.wrappedJSObject` to get access to the raw JS implementation.
+* **component**, **contract** [see section 'Replacing XPCOM registration' below](#replacing-xpcom-registration).
+* **category** use [`nsICategoryManager`](https://developer.mozilla.org/en/XPCOM_Interface_Reference/nsICategoryManager) through an experiment
+* **content**, **skin**, **resource** it is no longer easily possible to assign chrome://- and resource://-URLs from add-ons. However, all files of an add-on are now available through moz-extension:// URLs by default. To generate such URLs, use [`browser.runtime.getURL`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/getURL) or `extension.baseURL` from an experiment. Some WebExtension APIs also accept relative paths and generate suitable URLs automatically.
+* **locale** localization for WebExtensions is handled using the [`i18n` API](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/i18n), which uses a single `messages.json` file to store translations.
+* **style** use an experiment to monitor open windows, and inject the style through that experiment
+* **overlay** [see section 'Replacing Overlays' below](#replacing-overlays).
+
+There are no direct equivalents to manifest flags, so add-ons now need to provide their own mechanisms to switch code or resources depending on the runtime environment. Relevant information is accessible through the [`runtime` API](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/).
+
+
+## Replacing Overlays
+
+If you used XUL overlays, they are now finally gone for good and you must find an alternative. While it is technically possible to write an XUL overlay loader experiment, it is a better idea to re-evaluate the purpose of the overlay:
```

Comment:
> While it is technically possible to write an XUL overlay loader experiment, it is a better idea ...

Is that a reaction to what I did with quicktext? :-) I created that overlay API as a proof-of-concept for addons, where the authors do not want to invest any time but would be a great loss if not avail in TB78. That will be the minimum effort conversion to keep them alive a few more years, before experiments get turned off.

Reply:
No, I was not aware of QuickText back then. Though I admit that had to smirk when I looked at QuickText later on :P
Given that not having an overlay loader API in Thunderbird seems like a deliberate choice, I did not want to divert from the 'official' path here.

I have to disagree on the 'experiments get turned off' thing, though. I just don't see the Thunderbird devs to intentionally disable a useful feature for some pseudo-security – but I would expect more prominent warnings and/or an explicit opt-in requirement for the user (and likely the administrator in an enterprise context).